File: laba4/ex3.cpp
#include <iostream>
#include <vector>
#include <unordered_map>

using namespace std;

void generateSequence(long long x0, long long a, long long b, long long c) {
    vector<long long> sequence;
    unordered_map<long long, size_t> seen;

    sequence.push_back(x0);
    seen[x0] = 0;

    size_t i = 1;
    const size_t maxIterations = 1000;
    size_t cycleStart = 0;
    size_t cycleLength = 0;
    bool cycleFound = false;

    while (i < maxIterations) {
        long long next = (a * sequence[i - 1] + b) % c;
        sequence.push_back(next);

        if (seen.count(next)) {
            cycleStart = seen[next];
            cycleLength = i - cycleStart;
            cycleFound = true;
            break;
        }

        seen[next] = i;
        ++i;
    }

    if (cycleFound) {
        size_t targetSize = cycleStart + cycleLength * 2;
        while (sequence.size() < targetSize && sequence.size() < maxIterations) {
            long long next = (a * sequence.back() + b) % c;
            sequence.push_back(next);
        }

        cout << "Сгенерированная последовательность: ";
        size_t idx = 0;
        for (long long value : sequence) {
            if (idx >= targetSize) break;
            cout << value << " ";
            idx++;
        }
        cout << endl;
        cout << cycleStart + cycleLength + 1 << endl;
    } else {
        cout << "Цикл не найден в пределах " << maxIterations << " итераций." << endl;
        cout << "Сгенерированная последовательность: ";
        for (long long value : sequence) {
            cout << value << " ";
        }
        cout << endl;
    }
}

int main() {
    long long x0, a, b, c;

    cout << "Введите начальное значение X0 (0 <= X0 < C): ";
    cin >> x0;
    cout << "Введите множитель A (0 <= A <= C): ";
    cin >> a;
    cout << "Введите инкремент B (0 <= B <= C): ";
    cin >> b;
    cout << "Введите модуль C (C > 0): ";
    cin >> c;

    if (c <= 0 || x0 < 0 || x0 >= c || a < 0 || a > c || b < 0 || b > c) {
        cout << "Некорректные входные данные!" << endl;
        return 1;
    }

    generateSequence(x0, a, b, c);

    return 0;
}

File: laba4/ex3.go
package main

import (
	"fmt"
	"os"
)

func generateSequence(x0, a, b, c int64) {
	sequence := []int64{x0}       // Хранит всю последовательность
	seen := make(map[int64]int)   // Хранит значение и его индекс
	seen[x0] = 0                  // Начальное значение

	var cycleStart, cycleLength int
	cycleFound := false
	maxIterations := 1000

	// Генерация последовательности
	for i := 1; i < maxIterations; i++ {
		next := (a*sequence[i-1] + b) % c
		sequence = append(sequence, next)

		// Проверка на повторение
		if idx, exists := seen[next]; exists {
			cycleStart = idx
			cycleLength = i - idx
			cycleFound = true
			break
		}
		seen[next] = i
	}

	// Обработка результатов
	if cycleFound {
		// Догенерация до двух полных циклов
		targetSize := cycleStart + cycleLength*2
		for len(sequence) < targetSize && len(sequence) < maxIterations {
			next := (a*sequence[len(sequence)-1] + b) % c
			sequence = append(sequence, next)
		}

		// Вывод последовательности
		fmt.Print("Сгенерированная последовательность: ")
		for i := 0; i < targetSize && i < len(sequence); i++ {
			fmt.Printf("%d ", sequence[i])
		}
		fmt.Println()

		// Индекс начала повторения (1-based)
		fmt.Printf("Индекс начала повторения: %d\n", cycleStart+cycleLength+1)
	} else {
		fmt.Printf("Цикл не найден за %d итераций.\n", maxIterations)
		fmt.Print("Сгенерированная последовательность: ")
		for _, v := range sequence {
			fmt.Printf("%d ", v)
		}
		fmt.Println()
	}
}

func main() {
	var x0, a, b, c int64

	// Ввод параметров
	fmt.Print("Введите начальное значение X0 (0 <= X0 < C): ")
	fmt.Scan(&x0)

	fmt.Print("Введите множитель A (0 <= A <= C): ")
	fmt.Scan(&a)

	fmt.Print("Введите инкремент B (0 <= B <= C): ")
	fmt.Scan(&b)

	fmt.Print("Введите модуль C (C > 0): ")
	fmt.Scan(&c)

	// Проверка корректности
	if c <= 0 || x0 < 0 || x0 >= c || a < 0 || a > c || b < 0 || b > c {
		fmt.Println("Ошибка: Некорректные входные данные!")
		os.Exit(1)
	}

	generateSequence(x0, a, b, c)
}
